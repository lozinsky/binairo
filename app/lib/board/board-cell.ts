export type BoardCellKind = 'fixed' | 'regular';

export type BoardCellKindRaw = 0 | 1;

export type BoardCellState = 'B' | 'E' | 'R';

export type BoardCellStateRaw = 0 | 1 | 2;

export function isBoardCellKindRaw(value: unknown): value is BoardCellKindRaw {
  return typeof value === 'number' && value in BOARD_CELL_KIND_BY_RAW;
}

export function isBoardCellStateRaw(value: unknown): value is BoardCellStateRaw {
  return typeof value === 'number' && value in BOARD_CELL_STATE_BY_RAW;
}

const NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellState>> = {
  B: 'E',
  E: 'R',
  R: 'B',
};

const BALANCE_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, number>> = {
  B: -1,
  E: 0,
  R: 1,
};

const RAW_BY_BOARD_CELL_KIND: Readonly<Record<BoardCellKind, BoardCellKindRaw>> = {
  fixed: 0,
  regular: 1,
};

const BOARD_CELL_KIND_BY_RAW: Readonly<Record<BoardCellKindRaw, BoardCellKind>> = {
  0: 'fixed',
  1: 'regular',
};

const RAW_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellStateRaw>> = {
  B: 0,
  E: 2,
  R: 1,
};

const BOARD_CELL_STATE_BY_RAW: Readonly<Record<BoardCellStateRaw, BoardCellState>> = {
  0: 'B',
  1: 'R',
  2: 'E',
};

export type BoardCellValue = readonly [kind: BoardCellKindRaw, state: BoardCellStateRaw];

export class BoardCell {
  get balance() {
    return BALANCE_BY_BOARD_CELL_STATE[this.#state];
  }

  get isEmpty() {
    return this.#state === 'E';
  }

  get isFilled() {
    return !this.isEmpty;
  }

  get isFixed() {
    return this.#kind === 'fixed';
  }

  get state() {
    return this.#state;
  }

  readonly #kind: BoardCellKind;
  readonly #state: BoardCellState;

  constructor(kind: BoardCellKind, state: BoardCellState) {
    this.#kind = kind;
    this.#state = state;
  }

  static create(state: BoardCellState) {
    return new this('regular', state);
  }

  static from(value: BoardCellValue) {
    return new this(BOARD_CELL_KIND_BY_RAW[value[0]], BOARD_CELL_STATE_BY_RAW[value[1]]);
  }

  equals(other: BoardCell) {
    return this.#state === other.#state;
  }

  next() {
    return this.isFixed ? this : new BoardCell(this.#kind, NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE[this.#state]);
  }

  toFixed() {
    return new BoardCell('fixed', this.#state);
  }

  valueOf(): BoardCellValue {
    return [RAW_BY_BOARD_CELL_KIND[this.#kind], RAW_BY_BOARD_CELL_STATE[this.#state]];
  }
}

export function isBoardCellValue(value: unknown): value is BoardCellValue {
  if (Array.isArray(value) && value.length === 2) {
    return isBoardCellKindRaw(value[0]) && isBoardCellStateRaw(value[1]);
  }

  return false;
}
