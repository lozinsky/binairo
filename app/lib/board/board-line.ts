import { DoubleMatrixRotation, type MatrixLine, type MatrixRotatable } from '~/lib/matrix';
import { expectToBeDefined } from '~/shared/expect';

import { BoardCell, BoardCellState, type BoardCellValue } from './board-cell';

export type BoardLineValue = readonly BoardCellValue[];

export class BoardLine implements Iterable<BoardCell>, MatrixLine<BoardCell>, MatrixRotatable<BoardLine> {
  readonly #cells: readonly BoardCell[];

  constructor(cells: readonly BoardCell[]) {
    this.#cells = cells;
  }

  static blank(size: number) {
    return new this(Array.from({ length: size }, () => BoardCell.create(BoardCellState.E)));
  }

  static create(states: readonly BoardCellState[]) {
    return new this(states.map((state) => BoardCell.create(state)));
  }

  static from(value: BoardLineValue) {
    return new this(value.map((value) => BoardCell.from(value)));
  }

  [Symbol.iterator]() {
    return this.#cells[Symbol.iterator]();
  }

  at(index: number) {
    return this.#cells.at(index);
  }

  entries() {
    return this.#cells.entries();
  }

  equals(other: BoardLine) {
    if (this.length !== other.length) {
      return false;
    }

    return this.#cells.every((cell, index) => cell.equals(expectToBeDefined(other.at(index))));
  }

  includes(other: BoardLine) {
    if (this.length !== other.length) {
      return false;
    }

    return this.#cells.every((a, index) => {
      const b = expectToBeDefined(other.at(index));

      return b.isEmpty ? true : a.equals(b);
    });
  }

  indexOf(cell: BoardCell) {
    return this.#cells.indexOf(cell);
  }

  keys() {
    return this.#cells.keys();
  }

  lengthOf(state: BoardCellState) {
    return this.#cells.reduce((length, cell) => (cell.state === state ? length + 1 : length), 0);
  }

  rotate() {
    return new BoardLine(this.#cells.toReversed());
  }

  slice(start: number, end?: number) {
    return this.#cells.slice(start, end);
  }

  toDoubleRotation() {
    return DoubleMatrixRotation.run<BoardLine>(this);
  }

  valueOf(): BoardLineValue {
    return this.#cells.map((cell) => cell.valueOf());
  }

  get balance() {
    return this.#cells.reduce((balance, cell) => balance + cell.balance, 0);
  }

  get isEmpty() {
    return this.#cells.every((cell) => cell.isEmpty);
  }

  get isFilled() {
    return this.#cells.every((cell) => cell.isFilled);
  }

  get length() {
    return this.#cells.length;
  }

  get progress() {
    return this.#cells.reduce((progress, cell) => (cell.isFilled ? progress + 1 : progress), 0) / this.length;
  }
}
