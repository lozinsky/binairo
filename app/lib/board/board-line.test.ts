import { assert, expect, test } from 'vitest';

import { BoardCellKind, type BoardCellPair, BoardCellState } from './board-cell';
import { BoardLine } from './board-line';

test.each([4, 6, 8, 10])('returns blank board line', (size) => {
  expect(BoardLine.blank(size)).toMatchSnapshot();
});

test.each([
  [[BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]],
  [[BoardCellState.B, BoardCellState.R, BoardCellState.E]],
])('returns board from states', (states) => {
  expect(BoardLine.create(states)).toMatchSnapshot();
});

test.each([
  [
    [
      [BoardCellKind.Regular, BoardCellState.B] as BoardCellPair,
      [BoardCellKind.Regular, BoardCellState.B] as BoardCellPair,
      [BoardCellKind.Regular, BoardCellState.R] as BoardCellPair,
      [BoardCellKind.Regular, BoardCellState.E] as BoardCellPair,
    ],
  ],
  [
    [
      [BoardCellKind.Regular, BoardCellState.B] as BoardCellPair,
      [BoardCellKind.Regular, BoardCellState.R] as BoardCellPair,
      [BoardCellKind.Regular, BoardCellState.E] as BoardCellPair,
    ],
  ],
])('returns board from value of', (value) => {
  expect(BoardLine.from(value)).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns board line iterator', (line) => {
  expect(Array.from(line)).toMatchSnapshot();
});

test.each([
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), -1],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 0],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 1],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 2],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 3],
])('returns board cell at index', (line, index) => {
  expect(line.at(index)).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns board line entries', (line) => {
  expect(Array.from(line.entries())).toMatchSnapshot();
});

test.each([
  [
    BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
    BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  ],
  [
    BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
    BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  ],
])('returns "true" if one board line is equal to other', (line, other) => {
  expect(line.equals(other)).toMatchSnapshot();
});

test.each([
  [
    BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.E, BoardCellState.E]),
    BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  ],
  [
    BoardLine.create([BoardCellState.B, BoardCellState.E, BoardCellState.E]),
    BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  ],
])('returns "false" if one board line is not equal to other', (line, other) => {
  expect(line.equals(other)).toMatchSnapshot();
});

test.each([
  [
    BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
    BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.E, BoardCellState.E]),
  ],
  [
    BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
    BoardLine.create([BoardCellState.B, BoardCellState.E, BoardCellState.E]),
  ],
])('returns "true" if one board line is include other', (line, other) => {
  expect(line.includes(other)).toMatchSnapshot();
});

test.each([
  [
    BoardLine.create([BoardCellState.E, BoardCellState.R, BoardCellState.B, BoardCellState.B]),
    BoardLine.create([BoardCellState.R, BoardCellState.E, BoardCellState.B, BoardCellState.B]),
  ],
  [
    BoardLine.create([BoardCellState.E, BoardCellState.R, BoardCellState.R, BoardCellState.B]),
    BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.B, BoardCellState.B]),
  ],
  [
    BoardLine.create([BoardCellState.E, BoardCellState.R, BoardCellState.R, BoardCellState.B]),
    BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.B]),
  ],
])('returns "false" if one board line is not include other', (line, other) => {
  expect(line.includes(other)).toMatchSnapshot();
});

test.each([
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), -1],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 0],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 1],
  [BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]), 2],
])('returns board cell index of', (line, index) => {
  const cell = line.at(index);

  assert(cell !== undefined);

  expect(line.indexOf(cell)).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns board line keys', (line) => {
  expect(Array.from(line.keys())).toMatchSnapshot();
});

test.each([
  [BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]), BoardCellState.E],
  [BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]), BoardCellState.R],
  [BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]), BoardCellState.B],
])('returns length of board state in board line', (line, state) => {
  expect(line.lengthOf(state)).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('rotates board line', (line) => {
  expect(line.rotate()).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns board line value of', (line) => {
  expect(line.valueOf()).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E]),
  BoardLine.create([BoardCellState.R, BoardCellState.R, BoardCellState.B, BoardCellState.B]),
  BoardLine.create([BoardCellState.R, BoardCellState.R, BoardCellState.R, BoardCellState.B]),
  BoardLine.create([BoardCellState.R, BoardCellState.B, BoardCellState.B, BoardCellState.B]),
])('returns board line balance', (line) => {
  expect(line.balance).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E]),
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.E]),
])('returns "true" if board line is empty', (line) => {
  expect(line.isEmpty).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns "false" if board line is not empty', (line) => {
  expect(line.isEmpty).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.R]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.R]),
])('returns "true" if board line is filled', (line) => {
  expect(line.isFilled).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns "false" if board line is not filled', (line) => {
  expect(line.isFilled).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.B, BoardCellState.B, BoardCellState.R, BoardCellState.E]),
  BoardLine.create([BoardCellState.B, BoardCellState.R, BoardCellState.E]),
])('returns board line length', (line) => {
  expect(line.length).toMatchSnapshot();
});

test.each([
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.E, BoardCellState.E]),
  BoardLine.create([BoardCellState.E, BoardCellState.E, BoardCellState.B, BoardCellState.B]),
  BoardLine.create([BoardCellState.R, BoardCellState.R, BoardCellState.B, BoardCellState.B]),
])('returns board line progress', (line) => {
  expect(line.progress).toMatchSnapshot();
});
