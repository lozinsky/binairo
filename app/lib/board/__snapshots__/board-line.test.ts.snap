// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns "false" if board line is not empty 1`] = `false`;

exports[`returns "false" if board line is not empty 2`] = `false`;

exports[`returns "false" if board line is not filled 1`] = `false`;

exports[`returns "false" if board line is not filled 2`] = `false`;

exports[`returns "false" if one board line is not equal to other 1`] = `false`;

exports[`returns "false" if one board line is not equal to other 2`] = `false`;

exports[`returns "false" if one board line is not include other 1`] = `false`;

exports[`returns "false" if one board line is not include other 2`] = `false`;

exports[`returns "false" if one board line is not include other 3`] = `false`;

exports[`returns "true" if board line is empty 1`] = `true`;

exports[`returns "true" if board line is empty 2`] = `true`;

exports[`returns "true" if board line is filled 1`] = `true`;

exports[`returns "true" if board line is filled 2`] = `true`;

exports[`returns "true" if one board line is equal to other 1`] = `true`;

exports[`returns "true" if one board line is equal to other 2`] = `true`;

exports[`returns "true" if one board line is include other 1`] = `true`;

exports[`returns "true" if one board line is include other 2`] = `true`;

exports[`returns blank board line 1`] = `[⬜] [⬜] [⬜] [⬜]`;

exports[`returns blank board line 2`] = `[⬜] [⬜] [⬜] [⬜] [⬜] [⬜]`;

exports[`returns blank board line 3`] = `[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]`;

exports[`returns blank board line 4`] = `[⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜] [⬜]`;

exports[`returns board cell at index 1`] = `[⬜]`;

exports[`returns board cell at index 2`] = `[🟦]`;

exports[`returns board cell at index 3`] = `[🟥]`;

exports[`returns board cell at index 4`] = `[⬜]`;

exports[`returns board cell at index 5`] = `undefined`;

exports[`returns board cell index of 1`] = `2`;

exports[`returns board cell index of 2`] = `0`;

exports[`returns board cell index of 3`] = `1`;

exports[`returns board cell index of 4`] = `2`;

exports[`returns board from states 1`] = `[🟦] [🟦] [🟥] [⬜]`;

exports[`returns board from states 2`] = `[🟦] [🟥] [⬜]`;

exports[`returns board from value of 1`] = `[🟦] [🟦] [🟥] [⬜]`;

exports[`returns board from value of 2`] = `[🟦] [🟥] [⬜]`;

exports[`returns board line balance 1`] = `0`;

exports[`returns board line balance 2`] = `0`;

exports[`returns board line balance 3`] = `2`;

exports[`returns board line balance 4`] = `-2`;

exports[`returns board line entries 1`] = `
[
  [
    0,
    [🟦],
  ],
  [
    1,
    [🟦],
  ],
  [
    2,
    [🟥],
  ],
  [
    3,
    [⬜],
  ],
]
`;

exports[`returns board line entries 2`] = `
[
  [
    0,
    [🟦],
  ],
  [
    1,
    [🟥],
  ],
  [
    2,
    [⬜],
  ],
]
`;

exports[`returns board line iterator 1`] = `
[
  [🟦],
  [🟦],
  [🟥],
  [⬜],
]
`;

exports[`returns board line iterator 2`] = `
[
  [🟦],
  [🟥],
  [⬜],
]
`;

exports[`returns board line keys 1`] = `
[
  0,
  1,
  2,
  3,
]
`;

exports[`returns board line keys 2`] = `
[
  0,
  1,
  2,
]
`;

exports[`returns board line length 1`] = `4`;

exports[`returns board line length 2`] = `3`;

exports[`returns board line progress 1`] = `0`;

exports[`returns board line progress 2`] = `0.5`;

exports[`returns board line progress 3`] = `1`;

exports[`returns board line value of 1`] = `
[
  [
    1,
    0,
  ],
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
]
`;

exports[`returns board line value of 2`] = `
[
  [
    1,
    0,
  ],
  [
    1,
    1,
  ],
  [
    1,
    2,
  ],
]
`;

exports[`returns length of board state in board line 1`] = `1`;

exports[`returns length of board state in board line 2`] = `1`;

exports[`returns length of board state in board line 3`] = `2`;

exports[`rotates board line 1`] = `[⬜] [🟥] [🟦] [🟦]`;

exports[`rotates board line 2`] = `[⬜] [🟥] [🟦]`;
